sql/lease.go

```diff
@@ -0,0 +1,508 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Peter Mattis (peter@cockroachlabs.com)
+
+package sql
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"math/rand"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/cockroachdb/cockroach/client"
+	"github.com/cockroachdb/cockroach/security"
+	"github.com/cockroachdb/cockroach/sql/parser"
+	"github.com/cockroachdb/cockroach/util"
+	"github.com/cockroachdb/cockroach/util/hlc"
+	"github.com/cockroachdb/cockroach/util/log"
+	"github.com/cockroachdb/cockroach/util/retry"
+	"github.com/gogo/protobuf/proto"
+)
+
+// TODO(pmattis): Periodically renew leases for tables that were used recently and
+// for which the lease will expire soon.
+
+const (
+	leaseDuration    = int64(5 * time.Minute)
+	minLeaseDuration = int64(time.Minute)
+)
+
+var (
+	errLeaseVersionChanged = errors.New("lease version changed")
+)
+
+func nanosToDTimestamp(nanos int64) parser.DTimestamp {
+	return parser.DTimestamp{Time: time.Unix(0, nanos)}
+}
+
+// LeaseState holds the state for a lease. Exported only for testing.
+type LeaseState struct {
+	TableDescriptor
+	expiration int64
```

Comment:
there's something of a fidelity loss here with int64 in place of duration. it looks like it's just to make touching WallTime more ergonomic, but I think we can use Duration and not suffer too bad since the conversions will happen at the edge (and you can use `.GoTime()` to make handling time/Duration arithmetic more obvious).